### ISSUE TITLE
headStart is too deep when using struct arrays as system parameters

### ISSUE THREAD

**Issue by 0xhank** [latticexyz/mud#307]
TLDR: `abi.decode` has stack overflow errors when the data it decodes is too complex (even though the data isn't really that complex). I would recommend either setting the optimizer to false by default or providing an alternate method of execution that doesn't involve `abi.decode` when passing in complex structs.

In one of my systems, I am passing in an array of `Action` structs as follows:
```
    Action[] memory actions = abi.decode(arguments, (Action[]));
```
The struct is organized as follows:
```
struct Action {
  uint256 shipEntity;
  bytes[2] metadata;
  ActionType[2] actionType;
}
```

This gives me the following error:
CompilerError: Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables. When compiling inline assembly: Variable headStart is 1 slot(s) too deep inside the stack.

To fix it, I set `optimizer` to `false` in `foundry.toml`.

IMO there should be a workaround when you want to use complex structs as system parameters or the optimizer should be off by default

**Comment by alvrs** [comment]
I'm assuming `headStart` is another variable in your code?

I remember @cha0sg0d had a similar issue with a abi decoding of a complex struct - did you ever find a workaround?

**Comment by 0xhank** [comment]
headStart is baked into solidity, and appears to be used to track the start of the decoding stack? Related issue:
https://github.com/ethereum/solidity/issues/11638

**Comment by 0xhank** [comment]
I commented out the entire system other than `abi.decode` and the issue persisted

**Comment by cha0sg0d** [comment]
Couple of thoughts for @0xhank:

1. I got a `stack too deep` error when copying a large struct from memory to storage: https://github.com/cha0sg0d/solidity-struct-copy.

2. Have you tried reproducing this in a single test file? Might be easier to isolate the creation of the `Action` array and the subsequent encode / decode. My intuition is that `abi.encode / decode` of a dynamic array of structs might be creating a bunch of stack variables

3. Did you try compiling with `via_ir = true`?

**Comment by holic** [comment]
closing as I think this is out of scope of MUD
